### DISCUSSION TITLE
Exclude domains from cache

### DISCUSSION THREAD

**Discussion by c-michail** [TechnitiumSoftware/DnsServer#1026] [Ideas]
Hello,
  It would be nice to have a list to exclude eg. dynamic domains from caching. (especially with low TTL) so it will never serve stale data.

Thank you

**Comment by ShreyasZare** [comment]
Thanks for the post. There is no benefit from having such a feature even for dynamic domain names. The Serve Stale feature will also have no impact since that feature works only when the DNS server is unable to fetch updated info from upstream. The stale answer already has low TTL value of 30 sec by default which you can configure in Settings to even lower value.

Let me know if you still have concern regarding this.

**Reply by c-michail** [reply]
Let's say you have a domain: mydomain.dyndns.org with TTL 60 seconds.

Now the mydomain.dyndns.org IP changes in 10 seconds but Technitium will server will serve the old IP from the cache until it expires and gets the new IP from the upstream.

I guess having a list to bypass the cache and always query from the upstream for specific domains will always return the "updated IP".

**Reply by ShreyasZare** [reply]
The problem of having such a feature is that the DNS server will have to fetch the domain name for every call. This means it will be performing recursive resolution each time and asking those name servers for the domain name frequently. This means, it may take lot of time for the request to resolve and you may see frequent ServerFailure responses. Other issue is that a lot of name servers have query rate limiting configured and querying them frequently will cause them to stop answering any requests till the query rate goes down.

Even if you ignore all these issues, the other problem is that DNS response gets cached at the OS level, and at the web browser level too. Sure, the DNS server will not cache but there is no option to stop your OS from caching it and there is no such option with web browser either. So, you will end up running "ipconfig /flushdns" on your Windows and also going in advanced web browser settings and clear its DNS cache which will cause losing cache for all other the domain names and will be inconvenient too.

The domain already has a low 60 sec TTL which is ok for dynamic DNS usage. This is a tradeoff with having dynamic IP address with some inconvenience for a minute when the IP changes. The better way I would recommend is to ask your ISP to provide static IP address which most ISPs provide for a small additional fee. This is what I too have with my broadband setup.

**Reply by c-michail** [reply]
> The problem of having such a feature is that the DNS server will have to fetch the domain name for every call.

It depends on the application you run on the dynamic dns ip. In my application case i request it once to the the IP and only requesting it again in case of disconnection.

> Other issue is that a lot of name servers have query rate limiting

Yes i understand this that's why you don't setup eg. a HTTP Server over a dynamic ip.

> Sure, the DNS server will not cache but there is no option to stop your OS from caching it and there is no such option 
> with web browser either 

You can disable caching in windows and firefox (i dunno about chrome)

Windows: 
HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\Dnscache\Parameters
MaxCacheTtl = 0
MaxNegativeCacheTtl = 0

Firefox:
network.dnsCacheEntries = 0
network.dnsCacheExpiration = 0
network.dnsCacheExpirationGracePeriod = 0

Tested the above and indeed disable caching

Here's the story. According to the article below you can cache with minimum TTL 30-40mins almost everything.
[stop-using-ridiculously-low-dns-ttls](https://blog.apnic.net/2019/11/12/stop-using-ridiculously-low-dns-ttls/)

I currently use https://github.com/mokeyish/smartdns-rs to test that theory and it works without problems except the dynamic dns. 
Smartdns has an option to define custom TTL on a domain so i can apply 60s TTL to the dynamic dns only dispite the global rule for every domain to have 40m TTL.

So i wanted to archive the same with Technitium and skip using additional proxy that's why i requested this feature.
If i can manually configure the TTL per domain on the cache/general it's the same so i can have 40m TTL for everything and 60s TTL for dynamic dns.

Thank you

**Reply by ShreyasZare** [reply]
> It depends on the application you run on the dynamic dns ip. In my application case i request it once to the the IP and only requesting it again in case of disconnection.

> Yes i understand this that's why you don't setup eg. a HTTP Server over a dynamic ip.

I guess you did not get it. The client that will be resolving your domain name will cause the DNS server to fetch it from the DDNS service for every call and they may have rate limiting for DNS in place causing the resolution to fail. Its not about what you have on the server side, its about client side domain resolution.

> You can disable caching in windows and firefox (i dunno about chrome)

This would cause all caching to be disabled so it will make your web browsing experience worse compared to the one-off chance where your dynamic IP changed after a few days and it took few seconds for the domain to update and resolve to the new IP address.

So even with a feature that allows you to set 0 TTL for a given domain name, the domain is still going to end up getting cached at OS or browser level.

> Here's the story. According to the article below you can cache with minimum TTL 30-40mins almost everything.
[stop-using-ridiculously-low-dns-ttls](https://blog.apnic.net/2019/11/12/stop-using-ridiculously-low-dns-ttls/)

The article is about not using lower values for TTL. Lower values of TTL cause frequent recursive resolution that adds latency. Its actually arguing opposite of what you are requesting (0 TTL) here.

If your dynamic DNS service allows you to configure lower TTL value then it would be a better option than having such a feature.

**Reply by c-michail** [reply]
Dear Zare,
  I think you got me wrong.
  I want to set every minimum TTL in the cache to eg. 30 minutes dispite their TTL and exclude only a couple domains from this rule.

Thank you

**Reply by ShreyasZare** [reply]
I think what you asked earlier was to have an option to exclude domain names from cache, i.e. to set TTL to 0 for such domain names. Changing minimum TTL to 30 mins for all other domain names came up after you read that article. 

That article is intended for zone administrators and recommending what they should set as TTL for records in their zones based on some assumptions and measurements instead of using very low TTL values. It does not say that end-users should do this. Its really a bad idea to override TTL value for a domain name this way and totally not recommended.

If you can describe your actual setup that you have with dynamic IP like, what services have you hosted? where is the DNS server located (in LAN or public server)? who are the intended users of this service? etc. Then I will explain in detail on how having such an option to exclude domain names from cache is of no use.

Don't get me wrong but if this was a useful option then it would have been implemented already or have been planned to be implemented.

**Reply by c-michail** [reply]
> I think what you asked earlier was to have an option to exclude domain names from cache, i.e. to set TTL to 0 for such domain names. 
> Changing minimum TTL to 30 mins for all other domain names came up after you read that article.

I had that article in my mind since the question but i expressed it wrong from my side. Sorry

> That article is intended for zone administrators and recommending what they should set as TTL for records in their zones 
> based on some assumptions and measurements instead of using very low TTL values. It does not say that end-users should 
> do this. Its really a bad idea to override TTL value for a domain name this way and totally not recommended.

He concludes:
Setting a minimum TTL of 40 minutes instead of 5 minutes is not going to prevent users from accessing the service. However, it will drastically reduce latency, and improve privacy (more queries = more tracking opportunities) and reliability by avoiding unneeded queries.

Sure, there are valid reasons to use low DNS TTLs. But not for 75% of the Internet to serve content that is mostly immutable, but pointless to cache. And if, for whatever reasons, you really need to use low DNS TTLs, also make sure that cache doesn’t work on your website either. For the very same reasons.

So all i am asking if you can add to technitium some features of smartdns so we can set global minimum TTL and exclude specific domains from the global rules like:
ttl for all resource records "rr-ttl: ttl for all records" and
domain-rules
as described in: https://github.com/pymumu/smartdns/blob/doc/en/docs/configuration.md

Thank you

**Reply by ShreyasZare** [reply]
You are not understanding the point of that article at all. Its asking zone administrators to use a reasonable TTL value. End users should not do any such things on their local DNS resolver.

> He concludes:
Setting a minimum TTL of 40 minutes instead of 5 minutes is not going to prevent users from accessing the service. However, it will drastically reduce latency, and improve privacy (more queries = more tracking opportunities) and reliability by avoiding unneeded queries.
>
> Sure, there are valid reasons to use low DNS TTLs. But not for 75% of the Internet to serve content that is mostly immutable, but pointless to cache. And if, for whatever reasons, you really need to use low DNS TTLs, also make sure that cache doesn’t work on your website either. For the very same reasons.

If a zone admin has set a TTL of 24 hrs, then it means that the admin is very sure that the domain's IP is not gonna change frequently and they want people to cache that record for 24 hrs. If you set your local DNS resolver to cap TTL to 30 mins then your DNS server will unnecessarily keep on refreshing the record every 30 mins which is totally not required.

I am again saying, that article is not for end users and you should not interpret any thing that the article says or concludes.

> So all i am asking if you can add to technitium some features of smartdns so we can set global minimum TTL and exclude specific domains from the global rules like:
ttl for all resource records "rr-ttl: ttl for all records" and
domain-rules
as described in: https://github.com/pymumu/smartdns/blob/doc/en/docs/configuration.md

Technitium DNS server already has options to set minimum and maximum TTL values in the Settings > Cache section. However, its not recommended to change the default values since its of no use for normal scenarios and will just cause frequent refreshing of records in cache. These options are intended to be used only for specific scenarios.

The option to exclude domain names from being cached wont be implemented since its of no use. If such an option was available and you configure to exclude a specific domain name then it wont have any effect since that domain is still going to get cached at OS level and web browser level. Turning off caching at OS and browser level, like you mentioned earlier, will have effect on other domain names too and your web browsing will have noticeable latency and failure issues so its not a solution.

**Comment by c-michail** [comment]
Thank you for your time

**Reply by ShreyasZare** [reply]
You're welcome. Hope that I was able to explain correctly.
